### PR TITLE
[Bugfix] Prevent infinite update loop caused by a synchronous update in a passive effect

### PR DESCRIPTION
**Pull request by acdlite**
## The bug

In 18, passive effects are flushed synchronously if they are the result of a synchronous update. We have a guard for infinite update loops that occur in the layout phase, but it doesn't currently work for synchronous updates from a passive effect, which means if this scenario happens, the main thread will lock up and crash the tab.

The reason this probably hasn't come up yet is because synchronous updates inside the passive effect phase are relatively rare: you either have to imperatively dispatch a discrete event, like `el.focus`, or you have to call `ReactDOM.flushSync`, which triggers a warning. (In general, updates inside a passive effect are not encouraged.)

I discovered this because `useSyncExternalStore` does sometimes trigger updates inside the passive effect phase.

## The fix

The way we detect a "nested update" is if there's synchronous work remaining at the end of the commit phase.
    
Currently this check happens before we synchronously flush the passive effects. I moved it to after the effects are fired, so that it detects whether synchronous work was scheduled in that phase.

### CODE REVIEW

File: packages/react-reconciler/src/ReactFiberWorkLoop.new.js [hunk 1]
```diff
@@ -1897,6 +1897,15 @@ function commitRootImpl(root, renderPriorityLevel) {
...
   // Check if there's remaining work on this root
+  // TODO: This is part of the `componentDidCatch` implementation. Its purpose
+  // is to detect whether something might have called setState inside
+  // `componentDidCatch`. The mechanism is known to be flawed because `setState`
+  // inside `componentDidCatch` is itself flawed — that's why we recommend
+  // `getDerivedStateFromError` instead. However, it could be improved by
+  // checking if remainingLanes includes Sync work, instead of whether there's
+  // any work remaining at all (which would also include stuff like Suspense
+  // retries or transitions). It's been like this for a while, though, so fixing
+  // it probably isn't that urgent.
```
Comment (acdlite) [inline packages/react-reconciler/src/ReactFiberWorkLoop.new.js:1908]: Unrelated to fix, just noticed it while I was here

**Review by gaearon — APPROVED** [review]
